sklearn/metrics/_classification.py

```diff
@@ -2504,3 +2504,163 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+    >>> calibration_error(y_true, y_pred, n_bins=5)
+    0.0
+    >>> calibration_error(
+    ...     y_true, y_pred, n_bins=5, norm="max"
+    ... )
+    0.0
+    >>> y_true = np.array([0, 0, 0, 0] + [1, 1, 1, 1])
+    >>> calibration_error(y_true, y_pred, n_bins=5)
+    0.25
+    >>> calibration_error(
+    ...     y_true, y_pred, n_bins=5, norm="max"
+    ... )
+    0.25
+    """
+    y_true = column_or_1d(y_true)
+    y_prob = column_or_1d(y_prob)
+    assert_all_finite(y_true)
+    assert_all_finite(y_prob)
+    check_consistent_length(y_true, y_prob, sample_weight)
+    if any(y_prob < 0) or any(y_prob > 1):
+        raise ValueError("y_prob has values outside of [0, 1] range")
+
+    labels = np.unique(y_true)
+    if len(labels) > 2:
+        raise ValueError("Only binary classification is supported. "
+                         "Provided labels %s." % labels)
+
+    if pos_label is None:
+        pos_label = y_true.max()
+    if pos_label not in labels:
+        raise ValueError("pos_label=%r is not a valid label: "
+                         "%r" % (pos_label, labels))
+    y_true = np.array(y_true == pos_label, int)
+
+    norm_options = ('l1', 'l2', 'max')
+    if norm not in norm_options:
+        raise ValueError(f'norm has to be one of {norm_options}, got: {norm}.')
+
+    remapping = np.argsort(y_prob)
+    y_true = y_true[remapping]
+    y_prob = y_prob[remapping]
+    if sample_weight is not None:
+        sample_weight = sample_weight[remapping]
+    else:
+        sample_weight = np.ones(y_true.shape[0])
+
+    n_bins = int(n_bins)
+    if strategy == 'quantile':
+        quantiles = np.percentile(y_prob, np.arange(0, 1, 1.0 / n_bins) * 100)
+    elif strategy == 'uniform':
+        quantiles = np.arange(0, 1, 1.0 / n_bins)
+    else:
+        raise ValueError(
+            f"Invalid entry to 'strategy' input. Strategy must be either "
+            f"'quantile' or 'uniform'. Got {strategy} instead."
+        )
+
+    threshold_indices = np.searchsorted(y_prob, quantiles).tolist()
+    threshold_indices.append(y_true.shape[0])
+    avg_pred_true = np.zeros(n_bins)
+    bin_centroid = np.zeros(n_bins)
+    delta_count = np.zeros(n_bins)
+    debias = np.zeros(n_bins)
+
+    loss = 0.
+    count = float(sample_weight.sum())
+    for i, i_start in enumerate(threshold_indices[:-1]):
+        i_end = threshold_indices[i + 1]
+        # ignore empty bins
+        if i_end == i_start:
+            continue
+        delta_count[i] = float(sample_weight[i_start:i_end].sum())
+        avg_pred_true[i] = (np.dot(y_true[i_start:i_end],
+                                   sample_weight[i_start:i_end])
+                            / delta_count[i])
+        bin_centroid[i] = (np.dot(y_prob[i_start:i_end],
+                                  sample_weight[i_start:i_end])
+                           / delta_count[i])
+        if norm == "l2" and reduce_bias:
+            delta_debias = (
```

Comment:
Is this implementation of the bias term correct? I believe it should be `(avg_pred_true[i] * (avg_pred_true[i] - 1) * delta_count[i])/ count * (delta_count[I] - 1)` going off the definition [here](https://arxiv.org/pdf/1909.10155.pdf). The `1/count` doesn't get distributed properly, I think. This will produce nans if there's only one element in a bin, so I'm not sure if this was changed for numerical stability reasons. The authors of the paper mention [this](https://github.com/p-lambda/verified_calibration/blob/66baa0d460a6992131927c5df19c9c037c174f04/calibration/utils.py#L344) in their implementation, so it seems like the "undefined if bin size is 1" issue is expected.